[4.0] Plugin to add "CAPTCHA - hCaptcha" functionality

pe7er opened the pull request:
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/28613
~~**Note:** due to a bug with ALL Captcha's in Joomla 4, you will need to manually install PR https://github.com/joomla/joomla-cms/pull/28785 in case it has NOT yet been merged into Joomla 4.~~ The mentioned Captcha issue has been solved & merged into Joomla. You can test this PR against Joomla 4.0-dev

### Summary of Changes
This plugin adds "hCaptcha" as alternative CAPTCHA method in Joomla forms.

### Testing Instructions
In Joomla 4 back-end:
- Install the "CAPTCHA - hCaptcha" plugin (via System > Install > Discover)
- Enable + **configure** the **hCaptcha** plugin: 
System > Plugins > open the hCaptcha plugin
  - Add a Site Key + Secret Key
  - Enable the plugin
- **Set** hCaptcha as **default CAPTCHA** plugin
System > Global Configuration > [Site] tab > Default Captcha: select "**CAPTCHA - hCaptcha**"
- **Create** a **contact form**
Components > Contacts > Contacts > New
  - Add "Test" Name + Email
- **Create menu item** for **contact form**
Menus > Main Menu > New > 
  - Add menu title "Contact form"
  - Select "Menu Item Type": Contacts > Single Contact
  - Select "Contact": the "Test" contact that you've just created

On the front-end: Click on the "Contact form" menu item that you've just created.
At the bottom you should have the CAPTCHA with the hCaptcha form.

pe7er committed:
Plugin to add  functionality

pe7er commented:
In case this PR will be added to the Joomla 4 core, then **we could re-use large parts** of the de-DE, es-ES, fr-FR, nl-NL and pl-PL **language files** from: 
https://github.com/pe7er/hCaptcha/tree/master/language

richard67 commented on plugins/captcha/hcaptcha/hcaptcha.php:
```diff
@@ -0,0 +1,150 @@
+		}
+
+		// Discard spam submissions
+		if ($spam)
```
Hmm, I don't get this. Variable `$spam` is initialized with `false`in line 115 above in the same function, and here it is checked for being true, but in-between nothing happens with it.

pe7er replied:
Good point. I thought that it might be needed because reCaptcha used it.
But it does not make sense. I'll remove it. Thanks Richard!

richard67 replied:
Other thing I saw: Your PR still uses the html helper and not the asset manager for adding the js. But this can maybe be fixed with another, future PR. But if in doubt, ping Fedik.

pe7er committed:
removed unnecessary variable + if statement

pe7er commented:
@richard67 Thanks for your review!
I've removed the unnecessary $spam variable + if statement.

Quy commented on administrator/language/en-GB/plg_captcha_hcaptcha.ini:
```diff
@@ -0,0 +1,32 @@
+; Params
+PLG_HCAPTCHA_PUBLIC_KEY="Site Key"
+PLG_HCAPTCHA_PUBLIC_KEY_DESC="Site Keys are used to uniquely identify your site"
+PLG_HCAPTCHA_PRIVATE_KEY="Secret key"
```
```suggestion
PLG_HCAPTCHA_PRIVATE_KEY="Secret Key"
```

richard67 commented:
@Fedik Should this new plugin finally use the web asset manager to load the (external) js from hCaptcha? Currently it uses the html helper.

If web asset manager: Can it be done with a future PR, too?

Razzo1987 commented:
Test on registration page:

![immagine](https://user-images.githubusercontent.com/906604/80285447-52264180-8725-11ea-8b4a-99500fbe0fa4.png)


![immagine](https://user-images.githubusercontent.com/906604/80285409-0378a780-8725-11ea-8969-6c4417b1e4e4.png)

Razzo1987 commented:
Same problem non contact form:
![immagine](https://user-images.githubusercontent.com/906604/80285520-e1cbf000-8725-11ea-92ce-1d6b9afff683.png)
![immagine](https://user-images.githubusercontent.com/906604/80285576-4a1ad180-8726-11ea-8146-13b2d36d94cf.png)

Quy commented on plugins/captcha/hcaptcha/hcaptcha.php:
```diff
@@ -0,0 +1,143 @@
+		$hCaptchaResponse = $input->get('h-captcha-response', '', 'cmd');
+
+		// Check for Private Key
+		if (empty($privatekey))
```
```suggestion
		if (empty($privateKey))
```

Quy commented on plugins/captcha/hcaptcha/hcaptcha.php:
```diff
@@ -0,0 +1,143 @@
+		}
+
+		// Check for IP
+		if (empty($remoteip))
```
```suggestion
		if (empty($remoteIp))
```

Fedik commented:
@richard67 it better if it will be done initialy, but can be another pull of course

There nothing complicated:

```php
$this->app->getDocument()->getWebAssetManager()
  ->registerAndUseScript('plg_captcha_hcaptcha', 'https://hcaptcha.com/1/api.js', [], ['defer' => true]);

```

wilsonge committed:
Update plugins/captcha/hcaptcha/hcaptcha.php

Co-Authored-By: Quy <quy@fluxbb.org>

wilsonge committed:
Update plugins/captcha/hcaptcha/hcaptcha.php

Co-Authored-By: Quy <quy@fluxbb.org>

wilsonge committed:
Update administrator/language/en-GB/plg_captcha_hcaptcha.ini

Co-Authored-By: Quy <quy@fluxbb.org>

Quy commented on plugins/captcha/hcaptcha/hcaptcha.xml:
```diff
@@ -0,0 +1,74 @@
+					label="PLG_HCAPTCHA_THEME"
+					description="PLG_HCAPTCHA_THEME_DESC"
+					default="light"
+					filter="string"
+				>
```
```suggestion
					validate="options"
					>
```

Quy commented on plugins/captcha/hcaptcha/hcaptcha.xml:
```diff
@@ -0,0 +1,74 @@
+					label="PLG_HCAPTCHA_SIZE"
+					description="PLG_HCAPTCHA_SIZE_DESC"
+					default="normal"
+					filter="string"
+				>
```
```suggestion
					validate="options"
					>
```

Quy commented on administrator/language/en-GB/plg_captcha_hcaptcha.ini:
```diff
@@ -0,0 +1,32 @@
+PLG_HCAPTCHA_PRIVATE_KEY_DESC="A secret Key is used to verify your hCaptcha account"
+
+PLG_HCAPTCHA_THEME="Theme"
+PLG_HCAPTCHA_THEME_DESC="Select a light or dark theme"
```
Remove description. It adds no additional information.

pe7er replied:
Ok, I'll do that. Thanks for your suggestion!

Quy commented on administrator/language/en-GB/plg_captcha_hcaptcha.ini:
```diff
@@ -0,0 +1,32 @@
+PLG_HCAPTCHA_THEME_DARK="Dark"
+
+PLG_HCAPTCHA_SIZE="Size"
+PLG_HCAPTCHA_SIZE_DESC="Select compact for a smaller footprint"
```
Remove description. It adds no additional information as the options are self explanatory.

pe7er replied:
done!

Razzo1987 commented:
Good!
![immagine](https://user-images.githubusercontent.com/906604/80302573-16ce5600-87ab-11ea-8942-f3c9d0a82592.png)

Also tested in contact form

Razzo1987 commented:
I have tested this item :white_check_mark: successfully on 3a84ac096e0b0fb6a0c3dbe38fcfd03dd17bc2de<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

pe7er committed:
Update plugins/captcha/hcaptcha/hcaptcha.xml

Co-Authored-By: Quy <quy@fluxbb.org>

pe7er committed:
Update plugins/captcha/hcaptcha/hcaptcha.xml

Co-Authored-By: Quy <quy@fluxbb.org>

pe7er committed:
removed unnecessary parameter description

pe7er commented:
I've changed ```filter="string"``` to ```validate="options"``` and removed an unnecessary parameter description. Thanks @Quy !

richard67 commented:
> @richard67 it better if it will be done initialy, but can be another pull of course
> 
> There nothing complicated:
> 
> ```
> $this->app->getDocument()->getWebAssetManager()
>   ->registerAndUseScript('plg_captcha_hcaptcha.api', 'https://hcaptcha.com/1/api.js', [], ['defer' => true]);
> ```

@pe7er Could you check @Fedik 's suggestion, too? Would be nice to have it in the new J4 way initially.

pe7er committed:
changed HTMLHelper for getWebAssetManager

pe7er commented:
I've changed the HTMLHelper for getWebAssetManager to include the external JavaScript
Thanks @richard67 & @Fedik

richard67 commented:
@Razzo1987 Could you repeat your test as soon as you can find some time? There have been some final changes meanwhile in this Pull Request. Thanks in advance.

Quy commented on plugins/captcha/hcaptcha/hcaptcha.xml:
```diff
@@ -0,0 +1,72 @@
+					default=""
+					required="true"
+					filter="string"
+					size="100"
+					class="input-xxlarge"
```
```suggestion
```

pe7er replied:
@Quy Thank you for your suggestion to remove the size + class from the back-end parameter form. 
Why do you think it's better to remove it?

pe7er replied:
I've removed it. There was no change in the size of the parameter fields.

Quy commented on plugins/captcha/hcaptcha/hcaptcha.xml:
```diff
@@ -0,0 +1,72 @@
+					default=""
+					required="true"
+					filter="string"
+					size="100"
+					class="input-xxlarge"
```
```suggestion
```

Quy commented:
Add to installation scripts to remove having to discover.
Add to update scripts.

Razzo1987 commented:
I'm testing with:
- Theme: dark
- Size: compact
![immagine](https://user-images.githubusercontent.com/906604/80398437-e7f1d600-88b7-11ea-80df-d7833671f1ad.png)

Is not compact, is square XD

![immagine](https://user-images.githubusercontent.com/906604/80398786-76feee00-88b8-11ea-83ff-95fe7743f6a7.png)

On contact form:
- Theme: dark
- Size: normal

![immagine](https://user-images.githubusercontent.com/906604/80399146-04dad900-88b9-11ea-8a8e-8f08f7f80cb6.png)

![immagine](https://user-images.githubusercontent.com/906604/80399176-0efcd780-88b9-11ea-8ebb-6d54e98ef5af.png)

Perfect!

Razzo1987 commented:
I have tested this item :white_check_mark: successfully on 4580380a63e45ae2e1dc067b155a441b1ab105c8<br /><br />Work! I'm not a robot :wedding: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

adj9 commented:
I have tested this item :white_check_mark: successfully on 4580380a63e45ae2e1dc067b155a441b1ab105c8<br /><br />I confirm

before
after<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

SharkyKZ commented on plugins/captcha/hcaptcha/hcaptcha.php:
```diff
@@ -0,0 +1,145 @@
+	 */
+	public function onPrivacyCollectAdminCapabilities()
+	{
+		$this->loadLanguage();
```
This is not needed if `$autoloadLanguage` class property is set to `true`.

pe7er replied:
ok, thanks!

SharkyKZ commented on administrator/language/en-GB/plg_captcha_hcaptcha.ini:
```diff
@@ -0,0 +1,30 @@
+PLG_CAPTCHA_HCAPTCHA_XML_DESCRIPTION="This CAPTCHA plugin uses the hCAPTCHA service to prevent spammers. To get a site and secret key for your domain, go to <a href=\"https://www.hcaptcha.com/\" target=\"_blank\">https://www.hcaptcha.com</a>. To use this for new account registration, go to Options in the User Manager and select CAPTCHA - hCAPTCHA as the CAPTCHA."
+
+; Params
+PLG_HCAPTCHA_PUBLIC_KEY="Site Key"
```
Please use `PLG_CAPTCHA_HCAPTCHA` prefix for all strings.

pe7er replied:
Thanks for this correction!

richard67 commented:
> Add to installation scripts to remove having to discover.
> Add update scripts.

@pe7er Let me know if I shall help with the SQL stuff. The SQL stuff is required in my opinion.

pe7er committed:
Changed LanguageTags + removed size/class fields + loadLanguage

pe7er commented:
I've made the changes that @Quy + @SharkyKZ recommended.

richard67 commented:
> I've made the changes that @Quy + @SharkyKZ recommended.

@pe7er Except of Quy's request for installation and update SQL.

pe7er commented:
@richard67 Do you mean the SQL installation stuff in 
installation/sql/mysql/base.sql
installation/sql/postgresql/base.sql

richard67 commented:
> @richard67 Do you mean the SQL installation stuff in
> installation/sql/mysql/base.sql
> installation/sql/postgresql/base.sql

@pe7er These it needs for new installation, insert into extensions table. In addition it needs for each database type a new update sql, e.g. `4.0.0-2020-04-28.sql`, with the same insert statement for the extensions table. You could look it up by searching fr such inserts in present update sql scripts and copy from the latest one.

pe7er committed:
Addeinstallation and update SQL

pe7er commented:
I've added the installation and update SQL
Thanks @richard67 + @Quy

Quy commented on administrator/language/en-GB/plg_captcha_hcaptcha.ini:
```diff
@@ -7,24 +7,24 @@ PLG_CAPTCHA_HCAPTCHA="CAPTCHA - hCaptcha"
-PLG_HCAPTCHA_ERROR_NO_PRIVATE_KEY="Private Key (Secret Key) is missing"
-PLG_HCAPTCHA_ERROR_NO_IP="Error: No IP address"
-PLG_HCAPTCHA_ERROR_INCORRECT_CAPTCHA="The CAPTCHA was incorrect"
+PLG_CAPTCHA_HCAPTCHA_ERROR_NO_PUBLIC_KEY="Public Key (Site Key) is missing"
```
Please update where these are used.

pe7er replied:
Oops, I missed those. Thanks @Quy !

pe7er committed:
Changed language tags to PLG_CAPTCHA_HCAPTCHA

pe7er commented:
Fixed the error messages with the new language prefix: PLG_CAPTCHA_HCAPTCHA

C-Lodder commented:
Will this not conflict with https://extensions.joomla.org/extension/access-a-security/hcaptcha?

E.g if:

- A user has a Joomla 3 site
- Installs this extension from JED
- Updates to J4

pe7er commented:
Good point!
I use the same code base + naming for this J4 version and that other J3 version. 
I suppose that it depends on how the pre-update checker in Joomla 3.10 will work.

Quy commented:
Please fix conflicts.

pe7er committed:
solved merge conflicts in mysql + postgresql base.sql

pe7er commented:
@Quy I've solved the merge conflicts in the base.sql of mysql and of postgresql.

zero-24 commented:
> I use the same code base + naming for this J4 version and that other J3 version.
> I suppose that it depends on how the pre-update checker in Joomla 3.10 will work.

What kind of things would you expect that pre-update checker does? We have to use a different name for the version in core else it is going to get conflicted with yours from the JED. This is the same I did for my lazyloading and my httpheaders plugin just because of that upgrade conflict.

zero-24 commented on plugins/captcha/hcaptcha/hcaptcha.php:
```diff
@@ -0,0 +1,143 @@
+			throw new \RuntimeException(Text::_('PLG_CAPTCHA_HCAPTCHA_ERROR_NO_IP'));
+		}
+
+		$verifyResponse = file_get_contents(
```
please apply https://github.com/pe7er/hCaptcha/pull/21 here too

pe7er committed:
Replaced file_get_contents with HttpFactory

pe7er commented:
Thanks @zero-24 

- I've replaced file_get_contents with HttpFactory,
- added some checks on response + $verifyResponse (which are also in this J3 version https://github.com/pe7er/hCaptcha )
- solved a deprecated warning

SharkyKZ commented on plugins/captcha/hcaptcha/hcaptcha.php:
```diff
@@ -0,0 +1,160 @@
+	/**
+	 * Calls an HTTP POST function to verify if the user's guess was correct
+	 *
+	 * @param   string  $code  Answer provided by user. Not needed for the Hcaptcha implementation
```
This should probably be used. See https://github.com/joomla/joomla-cms/pull/26882.

pe7er replied:
Thanks for your suggestion. However I do not know what you mean, and what I have to change to improve it.

SharkyKZ replied:
`$code` argument seems to be used by 3rd party extensions. So line 114 should use it if it was provided:

    $hCaptchaResponse = $code ?? $input->get('h-captcha-response', '', 'cmd');

pe7er replied:
Ok, now I understand what you mean. Cheers! Fixed!

pe7er committed:
use  for  if set

richard67 commented:
@Razzo1987 @adj9 Could you test again with the latest changes? Thanks in advance.

Razzo1987 commented:
Testing fail:

```
Error

Invalid response from hcaptcha.com

Invalid field: Captcha
```

richard67 commented:
I can confirm @Razzo1987 's finding, and unfortunately there is nothing in PHP error log.

SharkyKZ commented on plugins/captcha/hcaptcha/hcaptcha.php:
```diff
@@ -0,0 +1,160 @@
+			throw new \RuntimeException(Text::_('PLG_CAPTCHA_HCAPTCHA_ERROR_CANT_CONNECT_TO_HCAPTCHA_SERVERS'));
+		}
+
+		if ($verifyResponse->code !== 200 || empty($verifyResponse->body))
```
```suggestion
		if ($verifyResponse->code !== 200 || $verifyResponse->body === '')
```
This should solve the issues above. But maybe there's a more preferred way.

pe7er replied:
Thanks @SharkyKZ, I've tested your suggestion and it works ok.

richard67 commented:
@pe7er Could you check [the review comment above](https://github.com/joomla/joomla-cms/pull/28798#discussion_r426188180) and [the corresponding report of the failed tests](https://github.com/joomla/joomla-cms/pull/28798#issuecomment-629690428)? I'd really like to see this fixed and tested and merged.

pe7er commented:
Thank you @Razzo1987, I can reproduce your issue.

pe7er committed:
Update plugins/captcha/hcaptcha/hcaptcha.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

pe7er commented:
@richard67 I've tested @SharkyKZ improvement (Thanks!) and added it to this PR.

richard67 commented:
@pe7er Will test tomorrow. Thanks @SharkyKZ 👍 .

richard67 commented:
I have tested this item :white_check_mark: successfully on 9528f0a079211f198d6b70635d30c8bafc51b6bc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

richard67 commented:
@Razzo1987 Hi Luca, could you test it one more time? I think it's good now. Thanks in advance.

richard67 commented:
Drone failure seems not to be related to this PR.

Quy commented on plugins/captcha/hcaptcha/hcaptcha.xml:
```diff
@@ -0,0 +1,68 @@
+	<copyright>Copyright (C) 2005 - 2019 Open Source Matters. All rights reserved.</copyright>
+	<license>GNU General Public License version 2 or later; see LICENSE.txt</license>
+	<description>PLG_CAPTCHA_HCAPTCHA_XML_DESCRIPTION</description>
+	<scriptfile>script.php</scriptfile>
```
```suggestion
```

Quy commented on administrator/language/en-GB/plg_captcha_hcaptcha.ini:
```diff
@@ -0,0 +1,33 @@
+
+; Params
+PLG_CAPTCHA_HCAPTCHA_PUBLIC_KEY="Site Key"
+PLG_CAPTCHA_HCAPTCHA_PUBLIC_KEY_DESC="Site Keys are used to uniquely identify your site"
+PLG_CAPTCHA_HCAPTCHA_PRIVATE_KEY="Secret Key"
+PLG_CAPTCHA_HCAPTCHA_PRIVATE_KEY_DESC="A secret Key is used to verify your hCaptcha account"
```
```suggestion
PLG_CAPTCHA_HCAPTCHA_PUBLIC_KEY_DESC="Site Keys are used to uniquely identify your site."
PLG_CAPTCHA_HCAPTCHA_PRIVATE_KEY="Secret Key"
PLG_CAPTCHA_HCAPTCHA_PRIVATE_KEY_DESC="A secret Key is used to verify your hCaptcha account."
```

pe7er committed:
Update administrator/language/en-GB/plg_captcha_hcaptcha.ini

Co-authored-by: Quy <quy@fluxbb.org>

pe7er committed:
Update plugins/captcha/hcaptcha/hcaptcha.xml

Co-authored-by: Quy <quy@fluxbb.org>

pe7er commented:
Thanks for your two small fixes (language file + xml manifest) @Quy !

richard67 commented:
Add back RTC because last change was only language string changes and the removal of not existing installation script from the manifest XML.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

brianteeman commented:
Is there a reason this was not merged 4 months ago?

rdeutz commented:
@richard67 can you have a lock at the conflicts

richard67 commented:
@rdeutz It needs not only to fix the conflicts, it also needs to rename the schema update SQL scripts to they run when updating a previous 4.0 to the version which includes this PR, and it needs to adapt these SQL scripts to changes which have happened meanwhile in the core for the extensions table, e.g. the `custom_data`column which has been added back and needs to provided with a value (empty string) in the insert statement. So that's a bit more work. Not sure yet if I have a push permission to Peter's branch.

richard67 committed:
Rename update SQL scripts and add custom_data column

- Rename the update SQL so it will run when updating a previous 4.0.
- Add the custom_data column to the update statement so it doesn't fail.

richard67 commented:
I've renamed and adjusted the update SQL script. Now I will fix the conflicts in the base.sql. Please be patient, it is not trivial.

richard67 committed:
Merge remote-tracking branch 'upstream/4.0-dev' into plg_captcha_hcaptcha

richard67 commented:
@rdeutz Conflicts solved. Now let's wait if the system test pass, then we know the base.sql scripts are ok. Theoretically it would need new tests for update to see if my changes in the update SQL are right.

richard67 commented:
I've chosen yesterday's date for the update SQL because today's date is already used in one of my current PR's.

richard67 commented:
Drone failed in unit tests, very likely just the randomly ocurring timeout. Restarted drone.

richard67 commented:
@rdeutz Drone passed. All done.

pe7er commented:
The RTC label has been removed. 
What has to be done to get it to RTC (or preferably merged if the code is ok) again?

richard67 commented:
> The RTC label has been removed.

No idea. Maybe by accident? But the bot has added it back now, it seems. @alikon Any idea what happened?

alikon commented:
> Theoretically it would need new tests for update to see if my changes in the update SQL are right.

richard67 commented:
Back to pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

richard67 commented:
I'll try to find testers on weekend and try to find time for testing myself. Thanks all for being so patient.

opn365 commented:
I have tested this item :white_check_mark: successfully on 773295f1d246a80bb26ed93f42d440a225ab7549<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

blueforce commented:
I have tested this item :white_check_mark: successfully on 773295f1d246a80bb26ed93f42d440a225ab7549<br /><br />Work, without any problems
Tested with Mac OSX, Chrome 86.0.42.xx<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28798">issues.joomla.org/tracker/joomla-cms/28798</a>.</sub>

richard67 commented:
@wilsonge Please decide if you want to have this in 4.0. If you asked me, I would say: Yes.

zero-24 commented on plugins/captcha/hcaptcha/hcaptcha.xml:
```diff
@@ -0,0 +1,67 @@
+	<author>Joomla! Project</author>
+	<authorEmail>admin@joomla.org</authorEmail>
+	<authorUrl>www.joomla.org</authorUrl>
+	<copyright>Copyright (C) 2005 - 2019 Open Source Matters. All rights reserved.</copyright>
```
```suggestion
	<copyright>Copyright (C) 2005 - 2020 Open Source Matters. All rights reserved.</copyright>
```

;)

zero-24 commented:
Once the decision is made this should be added here too: https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Extension/ExtensionHelper.php#L176 ;)

richard67 commented:
> Once the decision is made this should be added here too: https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Extension/ExtensionHelper.php#L176 ;)

That should be made even before. @pe7er Can you do that? If not let me know and I'll do it for you in this PR here.

wilsonge commented:
So my personal opinion on this is that we don't need this in core - and it can just be a 3rd party plugin - I don't see that need in core. In terms of the cloudflare original we know that the criteria is `one million queries per month (or 1,000 API calls per second)` - I highly doubt there are that many Joomla sites that are close to that level of traffic.

I understand that some people feel sketchy about using Google products for whatever privacy reasons they have. But I think that's still probably the minority case.

Maybe I'd have a different decision if we were implementing the first captcha plugin for the core default. But I don't really see any need to change out an existing working implementation on either of the grounds above.

pe7er commented:
Thanks for a making a decision on this. 
No problem! I'll release it as a free non-core plugin instead.

PhilETaylor commented:
<img width="618" alt="Screenshot 2021-03-23 at 12 21 12" src="https://user-images.githubusercontent.com/400092/112145458-4ad81400-8bd2-11eb-99aa-bade3499a0b4.png">

brianteeman commented:
I would really like to see this in Joomla 5. Especially as the feeling against the google options seems to be growing even stronger

pe7er commented:
Thanks, I understand the decision to not add it. 
I've already released it as a free plugin 3 years ago (for J3 and J4) :
https://extensions.joomla.org/extension/hcaptcha/

brianteeman commented:
I think the reasons then are less valid now

richard67 commented:
As hcaptcha is also located in US, it is also not GDPR compliant because is transferred to US. So using hcaptcha also requires a consent by the user in the EU. But in opposite to Google they at least tell in their privacy statement they collect, process and provide to 3rd parties (among which there are again US companies).

brianteeman commented:
That's not a reason to not include it

richard67 commented:
> That's not a reason to not include it

Right. But its an information which could be relevant when making a decision.

richard67 commented:
My personal opinion is that if we include a 3rd party captcha plugin from Google, we can also include the hCaptcha, as that seems to be a big more privacy friendly than the Google one. Another way could be not to include any captcha at all in the core. So I would be ok with the hCaptcha but that's just my opinion as a user.

micker commented:
In some eu country hcaptcha isnt compilant for gdrp  did you see me open discussion ? Maybe an honeypot Can be good for starting

laoneo commented:
The past showed us that the core should not have any reference to a 3rd party service. One of the biggest issues we face is when they do some updates. It is really hard for us to follow because of the backwards compatibility problems. The core should offer an easy integration, but not deliver the implementation for externals services at the same time.

brianteeman commented:
@laoneo that's a reason to never use any code that we haven't written ourselves which is a NIH policy we abandoned about ten years ago.
Do I really have to list all the external 3rd party services that are included in core.
TinyMCE
Codemirror
Bootstrap
Fontawesome
Jquery
skipto
sa11y
...

laoneo commented:
I think I have to explain you first the difference between a service and a dependency. A dependency is something we deliver with the installable package which is not maintained by us. At the time we ship it, there is no license issue and it doesn't call any external web site. So when we ship bootstrap in core and bootstrap.org is down or when they change their license it doesn't affect the CMS. We do not call any of their websites during productive mode, so there is no privacy issue at all.

With services it is completely different. Services do call external web sites while using in production mode. Does a service go down, change their license or removes a API endpoint we are using, then the CMS is broken. I do experience the same at the moment where zoom did break a deprecated API 4 months before it ends from one day to another. The worst use case I can think of is when a service is changing their license and the CMS has to pay per installation.

So there is a huge difference between a service and and a dependency. What the core should provide is an easy API that extension developers can integrate external services easily into core. But the core itself should not call any other website except the current one.

brianteeman commented:
Then you should have already removed all existing 3rd party services

micker commented:
https://github.com/joomla/joomla-cms/discussions/39662

laoneo commented:
@micker thanks for the link.
